modules/docs/resources/markdown/04-codegen/02-unions.md

```diff
@@ -149,6 +149,41 @@ are  encoded as such
 { "tpe": "second", "int": 42 }
 ```
 
+#### Open unions
+
+By default, `union` shapes in Smithy4s are considered closed.
+
+This means that any tags/discriminators that are not listed in the union's Smithy definition, will not be accepted as valid values in deserialization. As such, adding/removing values to such a union definition is a breaking change.
+
+This behavior is at odds with the current [specification of Smithy](https://smithy.io/2.0/guides/evolving-models.html#updating-unions) as of April 2025:
+
+> Unions in Smithy are considered "open"; it is a backward-compatible change to add new members to a union
```

Comment:
I don't like this way of introducing it : this bit of specification is borderline nonsensical. 

What this PR solves is a **PROTOCOL-SPECIFIC** problem, not a problem with how we treat the generated code. As a matter of fact, talking about backward/forward compatibility is very much a protocol-specific notion. 

I think this bit of documentation that talks about how unions are serialised should generally go to a JSON specific section of the docs, but that's a problem for another time. What I'm saying is that you shouldn't tie this PR back to what AWS states about union backward compat.

Reply:
> you shouldn't tie this PR back to what AWS states about union backward compat.

I still think it's worth a mention/link in codegen docs somewhere. To us it's a protocol-specific concern, but the spec is talking about model evolution in general, i.e. doesn't quantify which protocols it applies to. Someone with past Smithy exposure will be surprised to see we handle them differently by default

Reply:
Sure, but then how does AWS address this notion of open-unions in their own code-generators ? Cause that's what we should point out when we talk about this

Reply:
smithy-java generates an `$Unknown` member by default:

<img width="801" alt="image" src="https://github.com/user-attachments/assets/42bce3a9-e839-418d-962d-b81f77cc0b12" />

Reply:
what's interesting is what's in the body of `$UnknownMember`, I wish you had screenshotted that 😅 . 

I'm gonna go with the assumption that they capture the JSON payload (like you do in this PR). The problem is that "open-unions" only make sense if code-generators are protocol specific. 

As long as you add that the wording of this bit of doc (ie open-unions don't make sense in protocol-agnostic solutions), I'm fine with keeping these docs here until we move them elsewhere. 

It means changing the `by default, `union` shapes in Smithy4s are considered closed.`, because "open" and "closed" are nonsensical in protocol-agnostic solutions.

Reply:
since you asked...

<img width="1184" alt="image" src="https://github.com/user-attachments/assets/f47de8c1-5f78-49e7-9ec9-c4355856cf24" />

<img width="1020" alt="image" src="https://github.com/user-attachments/assets/9e3aa77d-06a4-4547-9bf3-0333f4caa766" />

it looks like they support decoding unknown members, but without keeping the actual structure that was decoded - just the member name. And you can't re-encode it back.

Reply:
tried to make it less controversial with ab5e6f9d7bf628e7e98e6228fcb7119443fe1cb6